src/details/ArborX_DetailsDistributor.hpp

```diff
@@ -31,6 +31,103 @@ namespace ArborX
+// (offsets).
+template <typename InputView, typename OutputView>
+static void
+determineBufferLayout(InputView batched_ranks, InputView batched_offsets,
+                      OutputView permutation_indices,
+                      std::vector<int> &unique_ranks, std::vector<int> &counts,
+                      std::vector<int> &offsets)
+{
+  auto const batched_offsets_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), batched_offsets);
+
+  auto const batched_ranks_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), batched_ranks);
+
+  ARBORX_ASSERT(unique_ranks.empty());
+  ARBORX_ASSERT(offsets.empty());
+  ARBORX_ASSERT(counts.empty());
+  ARBORX_ASSERT(permutation_indices.extent_int(0) == 0);
+  ARBORX_ASSERT(batched_ranks.size() + 1 == batched_offsets.size());
+  static_assert(
+      std::is_same<typename InputView::non_const_value_type, int>::value, "");
+  static_assert(std::is_same<typename OutputView::value_type, int>::value, "");
+
+  // In case all the batches are empty, return an empty list of unique_ranks and
+  // counts, but still have one element in offsets. This is conforming with
+  // creating the total offsets from batched_ranks and batched_offsets ignoring
+  // empty batches and calling sortAndDetermeineBufferLayout.
+  offsets.push_back(0);
+
+  auto const n = batched_ranks.size();
+  if (n == 0 || lastElement(batched_offsets) == 0)
+    return;
+
+  using DeviceType = typename InputView::traits::device_type;
+  using ExecutionSpace = typename InputView::traits::execution_space;
+
+  // Find the indices in batched_ranks for which the rank changes and store
+  // these ranks and the corresponding offsets in a new container that we can be
+  // sure to be large enough.
+  Kokkos::View<int *, DeviceType> compact_offsets(
+      Kokkos::ViewAllocateWithoutInitializing(batched_offsets.label()),
+      batched_offsets.size());
+  Kokkos::View<int *, DeviceType> compact_ranks(
+      Kokkos::ViewAllocateWithoutInitializing(batched_ranks.label()),
+      batched_ranks.size());
+
+  int n_unique_ranks;
+  Kokkos::parallel_scan(
+      ARBORX_MARK_REGION("compact_offsets_and_ranks"),
+      Kokkos::RangePolicy<ExecutionSpace>(0, n),
+      KOKKOS_LAMBDA(unsigned int i, int &update, bool last_pass) {
+        if ((i < batched_ranks.size() - 1 &&
+             batched_ranks(i + 1) != batched_ranks(i)) ||
+            i == batched_ranks.size() - 1)
+        {
+          if (last_pass)
+          {
+            compact_ranks(update) = batched_ranks(i);
+            compact_offsets(update + 1) = batched_offsets(i + 1);
+          }
+          ++update;
+        }
+      },
+      n_unique_ranks);
+
+  // Now create subviews containing the elements we actually need, copy
+  // everything to the CPU and fill the output variables.
+  auto restricted_offsets = InputView(
+      compact_offsets, std::make_pair(1, static_cast<int>(n_unique_ranks + 1)));
```

Comment:
Why this style and not `auto restricted_offsets = Kokkos::subview(compact_offsets, std::make_pair(...));`? And is there a reason to deal with subviews in the first place? Can't you just copy the correct things inside the `for` loop?

Reply:
I really don't care about the style to use for creating a subview. The documentation says this is slightly more efficient.

How else do I copy a subset of the elements from the GPU to the CPU?

Reply:
See the comment below about subset, and why I am not sure we need it.